crates/rpc-types/src/eth/transaction/request.rs

```diff
@@ -166,10 +148,327 @@ impl TransactionRequest {
+            to: self.to.expect("checked in will_build_4844"),
+            value: self.value.unwrap_or_default(),
+            access_list: self.access_list.unwrap_or_default(),
+            blob_versioned_hashes: self.blob_versioned_hashes.expect("checked in will_build_4844"),
+            max_fee_per_blob_gas: self.max_fee_per_blob_gas.expect("checked in will_build_4844"),
+            input: self.input.into_input().unwrap_or_default(),
+        }
+    }
+
+    fn check_reqd_fields(&self) -> Vec<&'static str> {
+        let mut missing = Vec::with_capacity(12);
+        if self.nonce.is_none() {
+            missing.push("nonce");
+        }
+        if self.gas.is_none() {
+            missing.push("gas_limit");
+        }
+        missing
+    }
+
+    fn check_legacy_fields(&self, missing: &mut Vec<&'static str>) {
+        if self.gas_price.is_none() {
+            missing.push("gas_price");
+        }
+    }
+
+    fn check_1559_fields(&self, missing: &mut Vec<&'static str>) {
+        if self.max_fee_per_gas.is_none() {
+            missing.push("max_fee_per_gas");
+        }
+        if self.max_priority_fee_per_gas.is_none() {
+            missing.push("max_priority_fee_per_gas");
+        }
+    }
+
+    /// Trim field conflicts, based on the preferred type
+    ///
+    /// This is used to ensure that the request will not be rejected by the
+    /// server due to conflicting keys, and should only be called before
+    /// submission via rpc.
+    pub fn trim_conflicting_keys(&mut self) {
+        match self.preferred_type() {
+            TxType::Legacy | TxType::Eip2930 => {
+                self.max_fee_per_gas = None;
+                self.max_priority_fee_per_gas = None;
+                self.max_fee_per_blob_gas = None;
+                self.access_list = None;
+                self.blob_versioned_hashes = None;
+                self.sidecar = None;
+            }
+            TxType::Eip1559 => {
+                self.gas_price = None;
+                self.access_list = None;
+                self.blob_versioned_hashes = None;
+                self.sidecar = None;
+            }
+            TxType::Eip4844 => {
+                self.gas_price = None;
+                self.access_list = None;
+            }
+        }
+    }
+
+    /// Check this builder's preferred type, based on the fields that are set.
+    ///
+    /// Types are preferred as follows:
+    /// - EIP-4844 if sidecar or max_blob_fee_per_gas is set
+    /// - EIP-2930 if access_list is set
+    /// - Legacy if gas_price is set and access_list is unset
+    /// - EIP-1559 in all other cases
+    pub fn preferred_type(&self) -> TxType {
+        if self.sidecar.is_some() || self.max_fee_per_blob_gas.is_some() {
+            TxType::Eip4844
+        } else if self.access_list.is_some() {
+            TxType::Eip2930
+        } else if self.gas_price.is_some() {
+            TxType::Legacy
+        } else {
+            TxType::Eip1559
+        }
```

Comment:
should 2930 only be if gas price is also set?

otherwise all txs with 1559 fee fields and an accesslist will be 2930

Reply:
my basic design approach here is to prefer the MOST specific (LEAST common) thing set, to try to conform to user intent in most cases. So preference order is basically as follows:

- Blobs are specific and critical to the tx, and have highest priority
- Access lists are specific and (almost always) not critical 
- legacy gas is specific and never critical
- eip1559 is the default case if no more-specific intention has been specified

> otherwise all txs with 1559 fee fields and an accesslist will be 2930

this seems like correct behavior to me, because if the user specified an access list it is a more-specific intention than specifying 1559 gas